tests/examples_test.go

```diff
@@ -0,0 +1,116 @@
+/*
+ * Copyright 2021 NMPolicy Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *	  http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package tests
+
+import (
+	"fmt"
+	"os"
+	"path/filepath"
+	"strings"
+	"testing"
+
+	assert "github.com/stretchr/testify/require"
+
+	"sigs.k8s.io/yaml"
+
+	"github.com/nmstate/nmpolicy/nmpolicy"
+	"github.com/nmstate/nmpolicy/nmpolicy/types"
+)
+
+const (
+	examplesPath = "../docs/examples"
+)
+
+type example struct {
+	name           string
+	policy         types.PolicySpec
+	currentState   []byte
+	generatedState []byte
+	capturedStates map[string]types.CaptureState
+}
+
+func (e *example) readFile(fileName string) ([]byte, error) {
+	return os.ReadFile(filepath.Join(examplesPath, e.name, fileName))
+}
+
+func (e *example) load() error {
+	policyMarshaled, err := e.readFile("policy.md")
+	if err != nil {
+		return err
+	}
+	policyMarshaled = convertToPolicyYAML(policyMarshaled)
+	if err = yaml.Unmarshal(policyMarshaled, &e.policy); err != nil {
+		return fmt.Errorf("failed unmarshaling example policy: %v", err)
+	}
+
+	e.currentState, err = e.readFile("current.yaml")
+	if err != nil {
+		return err
+	}
+
+	e.generatedState, err = e.readFile("generated.yaml")
+	if err != nil {
+		return err
+	}
+
+	capturedStatesMarshaled, err := e.readFile("captured.yaml")
+	if err != nil {
+		return err
+	}
+	e.capturedStates = map[string]types.CaptureState{}
+	if err := yaml.Unmarshal(capturedStatesMarshaled, &e.capturedStates); err != nil {
+		return fmt.Errorf("failed unmarshaling example captured states: %v", err)
+	}
+	return nil
+}
+
+func (e *example) run() (types.GeneratedState, error) {
+	obtained, err := nmpolicy.GenerateState(e.policy, e.currentState, types.CachedState{})
+	if err != nil {
+		return types.GeneratedState{}, err
+	}
+	return obtained, nil
+}
+
+func TestExamples(t *testing.T) {
+	examplesEntries, err := os.ReadDir(examplesPath)
+	assert.NoError(t, err)
+	for _, examplesEntry := range examplesEntries {
+		if !examplesEntry.IsDir() {
+			continue
+		}
+		e := example{name: examplesEntry.Name()}
+		t.Run(e.name, func(t *testing.T) {
+			err := e.load()
+			assert.NoError(t, err, "should successfully load the example")
```

Comment:
If it fails, don't we just to see why in the message? E.g.: "failed to load the example".
And if this is the case, this should be part of the returning error and not add here information... We do not really have any additional information to add at this abstraction level.

Reply:
This is not wrapping an error this is describing an assertion and that should be like that, I have forced the error so we see how it looks

```
--- FAIL: TestExamples (0.00s)
    --- FAIL: TestExamples/all-linux-bridges-down (0.00s)
        examples_test.go:100: 
            	Error Trace:	examples_test.go:100
            	Error:      	Received unexpected error:
            	            	forced error
            	Test:       	TestExamples/all-linux-bridges-down
            	Messages:   	should successfully load the example
```

Reply:
Well, you do not need to add text if the error tells you exactly what failed, right? It does not here, so that is the problem as I see it.
What the assertion does is self explanatory, we do not need to decorate each assertion with explanations, it will never end and has no value IMO.
There are rare cases where the assertion is not clear, i.e. one does not understand why the assertion is there. Then it makes sense to add information.

Not something that I intend to insist on anyway.